File: app/src/main/java/com/google/android/leanbacklauncher/notifications/PrescaledLayout.java
package com.google.android.leanbacklauncher.notifications;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.view.View.MeasureSpec;
import android.view.ViewGroup;
import com.google.android.leanbacklauncher.R;

public class PrescaledLayout extends ViewGroup {
    private View mContent;
    private float mScaleFactor;

    public PrescaledLayout(Context context) {
        super(context);
    }

    public PrescaledLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public PrescaledLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    protected void onFinishInflate() {
        super.onFinishInflate();
        setClipChildren(false);
        this.mScaleFactor = getResources().getFraction(R.fraction.lb_focus_zoom_factor_medium, 1, 1);
        this.mContent = getChildAt(0);
        this.mContent.setPivotX(0.0f);
        this.mContent.setPivotY(0.0f);
        this.mContent.setScaleX(1.0f / this.mScaleFactor);
        this.mContent.setScaleY(1.0f / this.mScaleFactor);
    }

    public int getContentHeight() {
        return (int) (((float) this.mContent.getMeasuredHeight()) / this.mScaleFactor);
    }

    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int width = MeasureSpec.getSize(widthMeasureSpec);
        this.mContent.measure(MeasureSpec.makeMeasureSpec((int) Math.ceil((double) (((float) width) * this.mScaleFactor)), 1073741824), 0);
        setMeasuredDimension(width, (int) (((float) this.mContent.getMeasuredHeight()) / this.mScaleFactor));
    }

    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        this.mContent.layout(0, 0, this.mContent.getMeasuredWidth(), this.mContent.getMeasuredHeight());
    }
}

File: app/src/main/java/com/amazon/tv/leanbacklauncher/notifications/PrescaledLayout.kt
package com.amazon.tv.leanbacklauncher.notifications

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import com.amazon.tv.leanbacklauncher.R
import kotlin.math.ceil

class PrescaledLayout : ViewGroup {
    private var mContent: View? = null
    private var mScaleFactor = 0f

    constructor(context: Context?) : super(context)
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    )

    override fun onFinishInflate() {
        super.onFinishInflate()
        clipChildren = false
        mScaleFactor = resources.getFraction(R.fraction.lb_focus_zoom_factor_medium, 1, 1)
        mContent = getChildAt(0)
        mContent?.pivotX = 0.0f
        mContent?.pivotY = 0.0f
        mContent?.scaleX = 1.0f / mScaleFactor
        mContent?.scaleY = 1.0f / mScaleFactor
    }

    val contentHeight: Int
        get() = (mContent!!.measuredHeight.toFloat() / mScaleFactor).toInt()

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val width = MeasureSpec.getSize(widthMeasureSpec)
        mContent!!.measure(
            MeasureSpec.makeMeasureSpec(
                ceil((width.toFloat() * mScaleFactor).toDouble()).toInt(), MeasureSpec.EXACTLY
            ), 0
        )
        setMeasuredDimension(
            width, (mContent!!.measuredHeight
                .toFloat() / mScaleFactor).toInt()
        )
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        mContent!!.layout(0, 0, mContent!!.measuredWidth, mContent!!.measuredHeight)
    }
}